### ISSUE TITLE
incorrect process history id in RunAuxiliary and LuminosityBlockAuxiliary ?

### ISSUE THREAD

**Issue by fwyzard** [cms-sw/cmssw#32596]

**Comment by fwyzard** [comment]
assign core

**Comment by fwyzard** [comment]
Please find attached a C++ module and a python configuration to reproduce the issue:
  - [HistoryAnalyzer.cc](https://github.com/cms-sw/cmssw/files/5765215/HistoryAnalyzer.cc.txt)
  - [testHistoryAnalyzer.py](https://github.com/cms-sw/cmssw/files/5765216/testHistoryAnalyzer.py.txt)

**Comment by makortel** [comment]
I **believe** the `ProcessHistoryID` of the Auxiliary object gets updated for the "input + current" (when the job produces any products) for Events in 
https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/FWCore/Framework/src/EventPrincipal.cc#L127
whereas for Lumis and Runs it is done in 
https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/IOPool/Output/src/RootOutputFile.cc#L455
https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/IOPool/Output/src/RootOutputFile.cc#L476

I'm wondering if this asymmetry is intentional or accidental.

**Comment by makortel** [comment]
@fwyzard Could you give some background for the question? Is this only a source of confusion, or do you have a use case that would require a change in behavior?

**Comment by dan131riley** [comment]
The EventAuxiliary branch can be fast-cloned, so it would seem that updating it would create an inconsistency between the fast-cloning vs updating cases.  Therefore, I would have guessed at the opposite of the behavior @fwyzard observes.

I'd also like to understand the use case, since I'm contemplating some changes to the handling of the EventAuxiliary branch (should reactivate #30949).

**Comment by fwyzard** [comment]
@makortel from a quick glance at the code, I think that what is missing for Run and Lumis is actually the equivalent of https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/FWCore/Framework/src/EventPrincipal.cc#L127 as you pointed out, but that the missing code should actually go somewhere in https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/FWCore/Framework/src/RunPrincipal.cc#L26-L34 and https://github.com/cms-sw/cmssw/blob/b1eb848eea0229b53887019d5c2953a289d68890/FWCore/Framework/src/LuminosityBlockPrincipal.cc#L16-L19 (or maybe in the constructors ?)

**Comment by makortel** [comment]
@fwyzard Indeed, but there is some code assuming that the Auxiliary objects of `LuminosityBlockPrincipal` and `RunPrincipal` have the `ProcessHistoryID` of the input, e.g.
https://github.com/cms-sw/cmssw/blob/64f0ebcc9683d5c5bfbae09e9832aa07c5c44d21/FWCore/Framework/src/EventProcessor.cc#L1610-L1613
and
https://github.com/cms-sw/cmssw/blob/64f0ebcc9683d5c5bfbae09e9832aa07c5c44d21/FWCore/Framework/src/InputSource.cc#L436
so the behavior seems to be intentional.

On the other hand, the behavior does not appear to be a fundamental property, so it might be possible to carefully change the `LuminosityBlockAuxiliary` and `RunAuxiliary`.

**Comment by makortel** [comment]
@dan131riley 
> The EventAuxiliary branch can be fast-cloned, so it would seem that updating it would create an inconsistency between the fast-cloning vs updating cases.

What are the conditions when `EventAuxiliary` can be fast-cloned?

**Comment by fwyzard** [comment]
Also, why would we keep unchanged the `EventAuxiliary` for runs and lumis, but change it for events ?

**Comment by makortel** [comment]
> Also, why would we keep unchanged the `EventAuxiliary` for runs and lumis, but change it for events ?

I wondered that myself as well, and digging from the history it was added in https://github.com/cms-sw/cmssw/commit/ca5d4be379ad6a03cc728246e254ecff466e829d, which unfortunately does not really help.

I went through all calls to `EventAuxiliary::processHistoryID()` that were found by DXR, and I did not see any calls that would use the `EventAuxiliary` from `EventPrincipal` or from `Event` (but I might have missed cases where the call to `EventPrincipal::aux()`/`Event::eventAuxiliary()` and `EventAuxiliary::processHistoryID()` are well separated). I did find the following assertion
https://github.com/cms-sw/cmssw/blob/40a52b912a834426e72ec8a38eed66938450e55a/IOPool/Output/src/RootOutputFile.cc#L402-L410
but I suspect the comment refers to lines 403,410 (and others) rather than the `assert()` itself.

**Comment by dan131riley** [comment]
@makortel I was misremembering, currently fast cloning of auxiliary branches is always disabled.

**Comment by makortel** [comment]
SubProcess appears to follow the same pattern, i.e. updating the `EventAuxiliary`'s `ProcessHistoryID` before processing the event
https://github.com/cms-sw/cmssw/blob/40a52b912a834426e72ec8a38eed66938450e55a/FWCore/Framework/src/SubProcess.cc#L406-L410
https://github.com/cms-sw/cmssw/blob/40a52b912a834426e72ec8a38eed66938450e55a/FWCore/Framework/src/SubProcess.cc#L423-L425
